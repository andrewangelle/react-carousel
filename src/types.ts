export interface PreloaderProps {
  currentIndex: number;
  autoPlay?: boolean;
  width: number;
  startTimer: () => void;
}

export interface PreloaderState {
  loading: boolean;
  done: boolean;
  loadCount: number;
}

export interface GlideState {
  currentIndex: number;
}

export interface GlideProps {
  autoPlay?: boolean;
  autoPlaySpeed?: number;
  dots?: boolean;
  height?: number;
  infinite?: boolean;
  width: number;
  onSlideChange?: () => void;
}
