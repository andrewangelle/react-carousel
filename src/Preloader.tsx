import React, { ReactElement } from 'react';

export interface PreloaderProps {
  images: string[];
}

export interface PreloaderState {
  loading: boolean;
  done: boolean;
}

export class Preloader extends React.Component<PreloaderProps, PreloaderState> {
  state: PreloaderState = {
    loading: true,
    done: false
  }

  componentDidMount() {
    this.preloadImages()
  }

  urls = () => {
    let urlResults: string[] = []

    React.Children.map(this.props.children,
      child => {
        const res = this.getImageUrls(child as ReactElement<any>)
        urlResults = [...urlResults, ...res]
      }
    )

    return urlResults
  }

  preloadImages() {
    const urls = [...this.props.images, ...this.urls()]
    let newImage
    let loadCount = 0

    const handleImageLoad = () => {
      loadCount = loadCount + 1

      if (loadCount === urls.length) {
        this.updateImageState();
      }
    }

    if (urls.length > 0) {
      urls.forEach(child => {
        console.log(child)
        newImage = new Image();
        newImage.onload = handleImageLoad;
        newImage.src = child;
      });
    }

    if (urls.length == 0) {
      this.updateImageState();
    }
  }

  getImageUrls(parent: ReactElement<any>) {
    const results: string[] = []
    const traverseTree = (object: any) => {
      if (object.type === 'img') {
        results.push(object.props.src)
      }
      if (object.props.children) {
        traverseTree(object.props.children)
      }
      return results
    }

    const res = traverseTree(parent)
    console.log({ res })
    return res
  }

  updateImageState() {
    this.setState({
      done: true,
      loading: false,
    });
  }
  render() {
    return (
      <>
        {this.state.loading && 'Loading ....'}
        {this.state.done && this.props.children}
      </>
    )
  }
}