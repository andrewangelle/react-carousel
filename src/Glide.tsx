import React, { ReactChild } from 'react';
import { Preloader } from './Preloader';
import './reactGlide.css';

export interface GlideProps {
  autoPlay?: boolean;
  autoPlaySpeed?: number;
  dots?: boolean;
  height?: number;
  infinite?: boolean;
  width: number;
  onSlideChange?: () => void;
}

export interface GlideState {
  loading: true;
  currentIndex: number;
  imagesLoaded: boolean;
}

class Glide extends React.Component<GlideProps, GlideState> {
  autoPlay: NodeJS.Timeout;

  state: GlideState = {
    loading: true,
    currentIndex: 0,
    imagesLoaded: false
  }

  componentWillUnMount() {
    clearInterval(this.props.autoPlaySpeed)
  }

  componentDidUpdate(_prevProps: GlideProps, prevState: GlideState) {
    const { currentIndex: prevIndex } = prevState;
    const { currentIndex } = this.state;
    const { onSlideChange } = this.props;

    if (currentIndex !== prevIndex && onSlideChange) {
      onSlideChange();
    }
  }

  startTimer = () => {
    if (this.props.autoPlay) {
      const { autoPlaySpeed = 5000 } = this.props
      this.autoPlay = setInterval(
        () => this.goToNextSlide(),
        autoPlaySpeed
      );
    }
  }

  goToSelectedDot = (index: number) => {
    this.setState({ currentIndex: index });
  }

  goToPrevSlide = () => {
    const { children } = this.props;
    const { currentIndex } = this.state;

    const lastSlide = React.Children.toArray(children).length - 1
    const nextIndex = currentIndex === 0 ? lastSlide : currentIndex - 1;

    this.setState({ currentIndex: nextIndex })
  }

  goToNextSlide = () => {
    const { children } = this.props;
    const { currentIndex } = this.state;

    const lastSlide = React.Children.toArray(children).length - 1
    const nextIndex = currentIndex === lastSlide ? 0 : currentIndex + 1;

    this.setState({ currentIndex: nextIndex })
  }

  render() {
    const { currentIndex } = this.state;

    const {
      infinite = false,
      children,
      dots = true
    } = this.props;

    const styleProps = {
      height: this.props.height,
      width: this.props.width,
    }

    return (
      <div className="glide--container" style={styleProps}>
        <Preloader
          startTimer={() => this.startTimer()}
          {...this.props}
          {...this.state}
        >
          {children}
        </Preloader>


        {(infinite || currentIndex !== 0) &&
          <button
            className="glide--prev-btn"
            onClick={() => {
              clearInterval(this.autoPlay)
              this.goToPrevSlide()
            }}
          >
            &#10094;
          </button>
        }

        {(infinite || currentIndex !== (children as ReactChild[]).length - 1) &&
          <button
            className="glide--next-btn"
            onClick={() => {
              clearInterval(this.autoPlay)
              this.goToNextSlide();
            }}
          >
            &#10095;
          </button>
        }

        {dots &&
          <section className="glide--dots" >
            {React.Children.map(children, (child, index) =>
              <span
                key={index}
                className={(currentIndex === index ? 'active-dot' : 'inactive-dot')}
                onClick={() => {
                  clearInterval(this.autoPlay)
                  this.goToSelectedDot(index)
                }}
              >
                &middot;
              </span>
            )}
          </section>
        }
      </div>
    );
  }
}

export { Glide }
