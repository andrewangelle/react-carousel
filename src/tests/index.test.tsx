import React from 'react';
import { shallow, mount } from './setupTests'

import { Glide, GlideState, GlideProps } from '../Glide';
import { Preloader } from '../Preloader';
import { PreloaderProps  } from '../types';

jest.useFakeTimers();

const props = {
  height: 500,
  width: 500,
  autoPlay: false,
  infinite: true,
  dots: true,
  onSlideChange: jest.fn()
}

const state = {
  currentIndex: 0,
  imagesLoaded: false
}

describe('Glide', () => {
  afterEach(() => jest.clearAllMocks())
  it('renders without crashing', () => {
    const baseProps = {
      width: 500,
      autoPlay: false,
      onSlideChange: jest.fn()
    }
    const component = shallow(
      <Glide {...baseProps}>
        <h1>Slide One</h1>
        <h1>Slide Two</h1>
        <h1>Slide Three</h1>
      </Glide>
    );

    expect(component.getElement()).toMatchSnapshot();
  });

  it('has children elements', () => {
    const component = shallow(
      <Glide  {...props} {...state}>
        <h1>Slide One</h1>
        <h1>Slide Two</h1>
        <h1>Slide Three</h1>
      </Glide>
    );
    expect(component.props().children.length).toBeGreaterThan(0)
  });

  it('renders first child element as first slide displayed', () => {
    const component = shallow(
      <Glide  {...props} {...state}>

        <h1>Slide One</h1>
        <h1>Slide Two</h1>
        <h1>Slide Three</h1>
      </Glide>
    );
    const element = component.instance().props.children![0].props.children;

    expect(element).toEqual('Slide One');
  });

  it('has width and height props', () => {
    const component = shallow(
      <Glide  {...props} {...state}>
        <h1>Slide One</h1>
        <h1>Slide Two</h1>
        <h1>Slide Three</h1>
      </Glide>
    );
    const widthProp = (component.instance().props as GlideProps).width
    const heightProp = (component.instance().props as GlideProps).height
    expect([widthProp, heightProp]).toBeTruthy();
  });

  it('sets container width according to prop', () => {
    const component = shallow(
      <Glide  {...props} {...state}>
        <h1>Slide One</h1>
        <h1>Slide Two</h1>
        <h1>Slide Three</h1>
      </Glide>
    );
    const userProp = (component.instance().props as GlideProps).width

    expect(userProp).toEqual(userProp);
  });

  it('changes to next index when next button is clicked', () => {
    const component = shallow(
      <Glide {...props}>
        <h1>Slide One</h1>
        <h1>Slide Two</h1>
        <h1>Slide Three</h1>
      </Glide>
    );

    const nextButton = component.find('button').last();
    expect((component.instance().state as GlideState).currentIndex).toEqual(0);

    nextButton.simulate('click');

    expect((component.instance().state as GlideState).currentIndex).toEqual(1);

    nextButton.simulate('click');
    expect((component.instance().state as GlideState).currentIndex).toEqual(2);

  });

  it.only('loops when next button is clicked', () => {
    const component = shallow(
      <Glide {...props}>
        <h1>Slide One</h1>
        <h1>Slide Two</h1>
        <h1>Slide Three</h1>
      </Glide>
    );

    const spy = jest.spyOn(component.instance() as Glide, 'onNextButtonClick')

    component.setState({ currentIndex: 2 })

    component
      .find('button')
      .last()
      .simulate('click');

    component.update()

    expect((component.instance().state as GlideState).currentIndex).toEqual(0);
    expect(spy).toHaveBeenCalled()
  });


  it('changes to previous index when prev button is clicked', () => {
    const component = shallow(
      <Glide  {...props}>
        <h1>Slide One</h1>
        <h1>Slide Two</h1>
        <h1>Slide Three</h1>
      </Glide>
    );
    const prevButton = component.find('button').first();
    expect((component.instance().state as GlideState).currentIndex).toEqual(0);

    prevButton.simulate('click');
    expect((component.instance().state as GlideState).currentIndex).toEqual(2);

    prevButton.simulate('click')
    expect((component.instance().state as GlideState).currentIndex).toEqual(1)

  });


  it('changes to next slide when next button is clicked', () => {
    const component = shallow(
      <Glide  {...props} {...state}>
        <h1>Slide One</h1>
        <h1>Slide Two</h1>
        <h1>Slide Three</h1>
      </Glide>
    );
    const nextButton = component.find('button').last();
    const elementOne = component.instance().props.children![0].props.children;
    const elementTwo = component.instance().props.children![1].props.children;
    const elementThree = component.instance().props.children![2].props.children;

    expect(elementOne).toEqual('Slide One');

    nextButton.simulate('click');
    expect(elementTwo).toEqual('Slide Two');

    nextButton.simulate('click');
    expect(elementThree).toEqual('Slide Three');

    nextButton.simulate('click');
    expect((component.instance().state as GlideState).currentIndex).toEqual(0)
  });

  it('changes to previous slide when prev button is clicked', () => {
    const component = shallow(
      <Glide {...props}>
        <h1>Slide One</h1>
        <h1>Slide Two</h1>
        <h1>Slide Three</h1>
      </Glide>
    );
    const prevButton = component.find('button').first();

    expect((component.find('Preloader').props() as PreloaderProps).currentIndex).toEqual(0)

    prevButton.simulate('click');
    expect((component.find('Preloader').props() as PreloaderProps).currentIndex).toEqual(2)

    prevButton.simulate('click');
    expect((component.find('Preloader').props() as PreloaderProps).currentIndex).toEqual(1)

  });


  it('changes slides when autoPlay is on', () => {
    const original = window.clearInterval;
    window.clearInterval = jest.fn();

    const component = mount(
      <Glide {...props} autoPlay={true} autoPlaySpeed={2000}>
        <h1>Slide One</h1>
        <h1>Slide Two</h1>
        <h1>Slide Three</h1>
      </Glide>
    );

    expect((component.instance().state as GlideState).currentIndex).toEqual(0);

    jest.runTimersToTime(4000);
    expect((component.instance().state as GlideState).currentIndex).toEqual(2)

    window.clearInterval = original
  });

  it('fires callback when when pagination is clicked', () => {
    const component = shallow(
      <Glide {...props}>
        <h1>Slide One</h1>
        <h1>Slide Two</h1>
        <h1>Slide Three</h1>
      </Glide>
    );

    expect(props.onSlideChange).not.toHaveBeenCalled()

    component
      .find('.inactive-dot')
      .first()
      .simulate('click')

    expect(props.onSlideChange).toHaveBeenCalled()
  })

  it('does not fire callback if index does not change', () => {
    const component = shallow(
      <Glide {...props}>
        <h1>Slide One</h1>
        <h1>Slide Two</h1>
        <h1>Slide Three</h1>
      </Glide>
    );

    component.setState({ currentIndex: 0 })

    expect(props.onSlideChange).not.toHaveBeenCalled();

  });

  it('does not fire callback if not provided', () => {
    const component = shallow(
      <Glide {...props} onSlideChange={undefined}>
        <h1>Slide One</h1>
        <h1>Slide Two</h1>
        <h1>Slide Three</h1>
      </Glide>
    );

    component.setState({ currentIndex: 1 })

    expect(props.onSlideChange).not.toHaveBeenCalled();
  });
});

describe('Preloader', () => {
  afterEach(() => jest.clearAllMocks())
  it('shows loader', () => {
    const wrapper = shallow(
      <Preloader
        currentIndex={0}
        width={500}
        autoPlay={false}
        startTimer={jest.fn()}
      >
        <img src="https://picsum.photos/700" />
        <h1>Slide Two</h1>
        <h1>Slide Three</h1>
      </Preloader>
    )

    expect(wrapper).toMatchSnapshot();
  });

  it('shows children', () => {
    const wrapper = shallow(
      <Preloader
        currentIndex={0}
        width={500}
        autoPlay={false}
        startTimer={jest.fn()}
      >
        <img src="https://picsum.photos/700" />
        <h1>Slide Two</h1>
        <h1>Slide Three</h1>
      </Preloader>
    )

    wrapper.setState({loading: false, done: true});
    wrapper.update()
    expect(wrapper).toMatchSnapshot();
  });
})
