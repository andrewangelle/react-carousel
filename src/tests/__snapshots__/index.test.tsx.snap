// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Glide renders without crashing 1`] = `
<div
  className="glide--container"
  style={
    Object {
      "width": 500,
    }
  }
>
  <Preloader
    autoPlay={false}
    currentIndex={0}
    imagesLoaded={false}
    loading={true}
    onSlideChange={[MockFunction]}
    startTimer={[Function]}
    width={500}
  >
    <h1>
      Slide One
    </h1>
    <h1>
      Slide Two
    </h1>
    <h1>
      Slide Three
    </h1>
  </Preloader>
  <button
    className="glide--next-btn"
    onClick={[Function]}
  >
    ❯
  </button>
  <section
    className="glide--dots"
  >
    <span
      className="active-dot"
      onClick={[Function]}
    >
      ·
    </span>
    <span
      className="inactive-dot"
      onClick={[Function]}
    >
      ·
    </span>
    <span
      className="inactive-dot"
      onClick={[Function]}
    >
      ·
    </span>
  </section>
</div>
`;

exports[`Preloader shows loader 1`] = `
<Preloader
  autoPlay={false}
  currentIndex={0}
  dots={true}
  imagesLoaded={false}
  infinite={true}
  loading={true}
  onSlideChange={[MockFunction]}
  startTimer={[MockFunction]}
  width={500}
>
  <LoadingSpinner
    width={500}
  >
    <div
      style={
        Object {
          "height": "378.7878787878788px",
          "width": "500px",
        }
      }
    >
      <div
        className="loading-indicator"
      />
    </div>
  </LoadingSpinner>
</Preloader>
`;
